Firestore/core/src/firebase/firestore/util/comparator_holder.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_COMPARATOR_HOLDER_H
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_COMPARATOR_HOLDER_H
+
+#include <type_traits>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * A holder of some comparator (e.g. std::less or util::Comparator) that
+ * implements the empty base optimization for the common case where the
+ * comparator occupies no storage.
+ */
+template <typename C, bool = std::is_empty<C>::value>
+class ComparatorHolder {
+ protected:
+  ComparatorHolder(const C& comparator) noexcept : comparator_(comparator) {
+  }
+
+  const C& comparator() const noexcept {
+    return comparator_;
+  }
+
+ private:
+  C comparator_;
+};
+
+// Implementation to use when C is empty.
+template <typename C>
+class ComparatorHolder<C, true> : private C {
+ protected:
+  ComparatorHolder(const C& comparator) noexcept : C(comparator) {
```

Comment:
Is the `: C(comparator)` part necessary if `C` is stateless?

Reply:
The alternative is to defeat a warning about an unused parameter.

Reply:
Entirely optional: add a short comment to that effect (e.g. `// This is just to avoid a warning about unused parameter`). Also, does compiler complain if the parameter has no name? (i.e., `ComparatorHolder(const C&) noexcept {}`?

Reply:
Removing the parameter name does the same thing less verbosely.